README.md

```diff
@@ -217,3 +217,45 @@ The resulting JSON structure derives the metric name and tag-map by decomposing
   }
 ]
 ```
+
+### Downsampling (a.k.a. roll-ups, normalized)
+
+Downsampling is the process of aggregating "raw" metrics, which are collected and conveyed at arbitrary timestamps, into deterministic time granularities, such as 5 minute and 1 hour. The intention is to keep aggregated metrics for much longer periods of time than the raw metrics. Downsampling also benefits queries with wider time ranges since it reduces the number of data points to be retrieved and rendered. 
+
+This process is also known as roll-up since it can be thought of finer grained data points rolling up into wider and wider grained levels of data. It is also referred to as normalized since the result of downsampling allows related metrics to be compared in time since the timestamps would align consistently even if those metrics were originally collected slightly "out of sync". 
+
+- Downsampling work is sliced into partitions, much like Cassandra's partitioning key concept
+- Partitions are an integer value in the range \[0, partition-count\)
+- A particular partition value is computed by hashing the tenant and series-set of a metric and then applycing a consistent hash, such as [the one provided by Guava](https://guava.dev/releases/21.0/api/docs/com/google/common/hash/Hashing.html#consistentHash-com.google.common.hash.HashCode-int-)
+- The application is configured with a "time slot width", which is used to define unit of tracking and the range queried for each downsample operation.
+- As a raw metric is ingested, the downsampling time slot is computed by rounding down the metric's timestamp to the next lowest multiple of the time slot width
+- The time slot width must be a common-multiple of the desired granularities. For example, if granularities of 5 minutes and 1 hour are used, then the time slot width must a multiple of an hour. With a one-hour time slot width, 12 5-minute downsamples would fit and one 1-hour downsample. 
+- The following table is used to track pending downsample sets by "upserting" the corresponding row as part of raw metric ingestion
+
+`pending_downsample_sets`:
+
+Type | Column
+-----|------------
+PK   | partition
+CK   | time_slot
+CK   | tenant
+CK   | series_set
+
+- The `time_slot` is the first of clustering keys to ensure that oldest raw data is processed before newer raw data
+- This application will be replicated/scaled-out in order to accommodate the workload of the series-set cardinality
+- Each replica is configured with a distinct set of partition values that it will process
+- On a periodic basis, each replica will query the `pending_downsample_sets` for each partition that it owns along with an upper bound on the `time_slot` to ensure that very recent metrics are allowed a certain amount of time to settle before being considered for downsampling
```

Comment:
Can you be a little more explicit about what happens on this periodic basis.  Let's assume 3 granularities, 5minutes, 1 hour and 4 hours.

how does it decide when to do each of those granularities?
how does it determine which granularities to do and which have already been done?
when it finishes all the granularities for a timeslot, does it indicate that in anyway, so it knows not to return to it?
describe what happens when a rollup node restarts
describe what happens when delayed metrics arrive after the corresponding rollup has been completed

Reply:
Yes, I later realized I had forgotten to mention the important part where the rows are deleted from `pending_downsample_sets` as they're processed. I'll add that in here and document what I'm thinking for the handling of late arrivals.

Reply:
I guess I'm still not following the periodic processing.  Lets say the granularities are 5 and 60 minutes.  So the ts width is 60.

At time 0 you start getting metrics, 5 minutes later you start processing the first 5min aggregation.  5 minutes after that do you reprocess that first 5 minutes or not?  How do you know whether it is completed or some delayed metrics have arrived?

Reply:
or do you mean no processing is done until the whole ts slot has passed?

Reply:
> or do you mean no processing is done until the whole ts slot has passed?
> 
> 

Yes, I was thinking a given time slot is not processed until at least that entire time slot has elapsed along with a configurable amount of grace period. I'll put together some kind of "timeline diagram" that will help me think through and depict the input/output processing of the slots.

Reply:
...and re-reading your first comment, granularities of 5m, 1h, 4h would dictate that the minimum time slot width is 4 hours. As such, if a metric arrived with a timestamp near the start of that 4 hour slot, then it would be about 4 hours and a bit before downsampling is even considered for that metric. That does start to push the limit of acceptability for availability of those downsampled values.